core/utils.py

```diff
@@ -1,183 +1,77 @@
-    return constant_step_size(priority_dir(dir, k), step_size)
-
-def priority_dir(dir, k=5):
-    dir_arry = dir.to_numpy()
-    sorted_idx = np.argsort(-np.abs(dir_arry[0,:]))
-    dir_new = np.zeros_like(dir_arry)
-    dir_new[:,sorted_idx[:k]] = dir_arry[:,sorted_idx[:k]]
-    #sparse_dir = dir.copy()
-    #sparse_dir.loc[:] = dir_new
-    return pd.DataFrame(columns=dir.columns, data=dir_new)
-
-def constant_step_size(dir, step_size=1):
-    normalization = np.linalg.norm(dir.to_numpy())
-    if normalization <= MIN_DIRECTION:
-        return dir
-    return (step_size*dir) / normalization
-
-def preference_dir(preferences, epsilon, max_step_size, dir):
-    for dimension in preferences:
-        if np.abs(dir[dimension]) > epsilon:
-            perturbed_dir = np.zeros_like(dir)
-            perturbed_dir[dimension] = min(dir[dimension], max_step_size)
-            return perturbed_dir
-    return np.zeros_like(dir)
+    new_direction = direction + noise
+    new_direction = (new_direction / np.linalg.norm(new_direction)) * norm
+    return new_direction
 
-def normal_alpha(dist, width=1):
-    return np.exp(-0.5 * (dist/width)**2)
 
-def volcano_alpha(dist, cutoff=0.5, degree=2):
-    return np.where(dist <= cutoff, 1/cutoff**degree, 1/dist**degree)
+def constant_step_size(
+    direction: recourse_adapter.EmbeddedSeries, step_size: float = 1
+) -> recourse_adapter.EmbeddedSeries:
+    """Rescales a vector to a given fixed size measured by L2 norm.
 
-def private_alpha(dist, cutoff=0.5, degree=2):
-    return 1/dist * np.where(dist <= cutoff, 1/cutoff**degree, 1/dist**degree)
-
-def model_early_stopping(model, point, cutoff=0.7):
-    _, pos_proba = model.predict_proba(point.to_numpy())[0]
-    return pos_proba >= cutoff
-
-def epsilon_compare(point, df):
-    numeric_columns = point.select_dtypes(include=np.number).columns
-    other_columns = point.columns.difference(numeric_columns)
-
-    numeric_diff = np.ones((df.shape[0], df.shape[1])).astype(np.bool8)
-    if len(numeric_columns) > 0:
-        numeric_diff = (np.abs(df[numeric_columns].to_numpy() - point[numeric_columns].to_numpy()) >= EQ_EPSILON)
-    other_diff = np.ones((df.shape[0], df.shape[1])).astype(np.bool8)
-    if len(other_columns) > 0:
-        other_diff = (df[other_columns].to_numpy() != point[other_columns].to_numpy())
+    Args:
+        direction: The vector to rescale.
+        step_size: The target L2 norm of the rescaled vector.
 
-    return numeric_diff & other_diff
+    Returns:
+        A new vector with direction equal to the original but rescaled to the
+        magnitude given by `step_size`.
+    """
+    normalization = np.linalg.norm(direction.to_numpy())
+    if normalization <= MIN_DIRECTION:
+        return direction
+    return (step_size * direction) / normalization
 
 
-def random_poi(dataset: pd.DataFrame, column = 'Y', label: Any = -1, drop_label: bool = True) -> pd.Series:
+def random_poi(
+    dataset: pd.DataFrame,
+    column: str = "Y",
```

Comment:
What does the "Y" represent here? Should this be "label" or some other indicator?

Reply:
Consider renaming this to `label_column_name` if this indicates the name of the dataset column that labels are contained in.

Reply:
Good point. Since I never rely on the default anyway here and or in the adapter classes, I made label_column a positional argument with no default.